sql/api/src/main/scala/org/apache/spark/sql/catalyst/parser/DataTypeAstBuilder.scala

```diff
@@ -45,6 +45,78 @@ class DataTypeAstBuilder extends SqlBaseParserBaseVisitor[AnyRef] {
     withOrigin(ctx)(StructType(visitColTypeList(ctx.colTypeList)))
   }
 
+  override def visitSingleRoutineParamList(ctx: SingleRoutineParamListContext): StructType = {
+    withOrigin(ctx)(StructType(visitSimplifiedColDefinitionList(ctx.colDefinitionList())))
+  }
+
+  private def visitSimplifiedColDefinitionList(
+      ctx: ColDefinitionListContext): Seq[StructField] = withOrigin(ctx) {
+    ctx.colDefinition().asScala.map(visitStructField).toSeq
+  }
+
+  /**
+   * Create a [[StructField]] from a column definition which allows options like COMMENT and
+   * DEFAULT.
+   *
+   * Don't handle generation expression since this function is currently only used for creating
+   * SQL functions which don't support generation expressions. The rejection logic is in
+   * [[SqlBaseParserVisitor#visitCreateUserDefinedFunction()]] implementation.
+   */
+  private def visitStructField(
```

Comment:
This method looks very similar to `visitColDefinition`. shall we invoke `visitColDefinition` here, fail if `ColumnDefinition#generationExpression/identityColumnSpec` is defined, and call `ColumnDefinition#toV1Column` to get `StructField`.

Reply:
done